### ISSUE TITLE
CUDA illegal memory access when using multiple losses and optimizers

### ISSUE THREAD

**Issue by akshaykvnit** [NVIDIA/apex#922]
I'm using 2 separate losses (loss0 and loss1) and 2 separate Adam optimizers (optim0 and optim1) but they share the same model. 

I tried (according to the [documentation](https://nvidia.github.io/apex/advanced.html#multiple-models-optimizers-losses))
```
....
model, [optim0, optim1] = amp.initialize(model, [optim0, optim1], opt_level='O1', num_losses=2)
....
optim0.zero_grad()
optim1.zero_grad()
# forward passes and loss calculations
....
with amp.scale_loss(loss0, [optim0, optim1], loss_id=0) as scaled_loss:
    scaled_loss.backward()
with amp.scale_loss(loss1, [optim0, optim1], loss_id=0) as scaled_loss:
    scaled_loss.backward()

optim0.step()
optim1.step()
```

However, this gives a CUDA illegal memory access. Then, I tried a few variations like
1. removing `num_losses=2` and `loss_id` parameters
2. adding `retain_graph=True` in the `backward` calls
3. giving only one `optim` to one scaled_loss at a time like
```
with amp.scale_loss(loss0, optim0, loss_id=0) as scaled_loss:
    scaled_loss.backward()
with amp.scale_loss(loss1, optim1, loss_id=0) as scaled_loss:
    scaled_loss.backward()
```
In these, (1) and (2) results in illegal memory access error, while (3) starts running for few iterations, but always gives gradient overflow (even though I used zero_grad at every iteration) and eventually results in illegal memory access error.

My usual use case (normal PyTorch) which works fine is
```
optim0.zero_grad()
optim1.zero_grad()
# forward passes and loss calculations
....

loss0.backward(retain_graph=True)
loss1.backward()

optim0.step()
optim1.step()
```

Adding the losses together and using a single backward call and a single optimizer might work, but since I use Adam optimizers, it will not be equivalent to using separate optimizers as Adam uses the previous iteration gradients.

**Comment by akshaykvnit** [comment]
It worked when I used PyTorch 1.6 AMP instead of apex.
